src/Pos/Block/Pure.hs

```diff
@@ -530,6 +554,22 @@ verifyBlock VerifyBlockParams {..} blk =
                  , sformat ("main block has unknown attributes: "%build) attrs)
                ]
 
+-- Type alias for common type used inside 'verifyBlocks'
+type VerifyBlocksIter ssc =
+    (Maybe SlotLeaders, Maybe ProxySKHeavyMap, Maybe (BlockHeader ssc), VerificationRes)
+
+-- Applies block certificates to 'ProxySKHeavyMap'. Used in
+-- 'verifyBlocks'. It's not the best place for it here, but i can't
+-- put it to "Delegation.Logic" because it leads to dependency cycle.
+pskHeavyMapApplyBlock :: MainBlock ssc -> ProxySKHeavyMap -> ProxySKHeavyMap
+pskHeavyMapApplyBlock block m = flip execState m $ do
+    let (toDelete,toReplace) =
+            partition (\ProxySecretKey{..} -> pskIssuerPk == pskDelegatePk)
+                      (view mainBlockDlgPayload block)
+    for_ toDelete $ \psk -> identity %= HM.delete (pskIssuerPk psk)
+    for_ toReplace $ \psk -> identity %= HM.insert (pskIssuerPk psk) psk
```

Comment:
I wrote in vatneek that I didn't have enough brain in the morning to understand whether logic you wrote is correct. Will do it soon and then will approve/reject.